xds/src/main/java/io/grpc/xds/RingHashLoadBalancer.java

```diff
@@ -0,0 +1,389 @@
+      EquivalentAddressGroup addrKey = entry.getKey();
+      double normalizedWeight = (double) entry.getValue() / totalWeight;
+      // TODO(chengyuanzhang): is using the list of socket address correct?
+      StringBuilder sb = new StringBuilder(addrKey.getAddresses().toString());
+      sb.append('_');
+      targetHashes += scale * normalizedWeight;
+      long i = 0L;
+      while (currentHashes < targetHashes) {
+        sb.append(i);
+        long hash = hashFunc.hashAsciiString(sb.toString());
+        ring.add(new RingEntry(hash, addrKey));
+        i++;
+        currentHashes++;
+        sb.deleteCharAt(sb.length() - 1);
+      }
+    }
+    Collections.sort(ring);
+
+    List<Subchannel> removedSubchannels = new ArrayList<>();
+    for (EquivalentAddressGroup addr : removedAddrs) {
+      removedSubchannels.add(subchannels.remove(addr));
+    }
+
+    // Update the picker before shutting down the subchannels, to reduce the chance of race
+    // between picking a subchannel and shutting it down.
+    updateBalancingState(new RingHashPicker(
+        Collections.unmodifiableList(ring), ImmutableMap.copyOf(subchannels)));
+    for (Subchannel subchann : removedSubchannels) {
+      shutdownSubchannel(subchann);
+    }
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    if (currentState != READY) {
+      helper.updateBalancingState(TRANSIENT_FAILURE, new ErrorPicker(error));
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    logger.log(XdsLogLevel.INFO, "Shutdown");
+    for (Subchannel subchannel : subchannels.values()) {
+      shutdownSubchannel(subchannel);
+    }
+  }
+
+  private void updateBalancingState(SubchannelPicker picker) {
+    checkState(!subchannels.isEmpty(), "no subchannel has been created");
+    ConnectivityState overallState = null;
+    for (Subchannel subchannel : subchannels.values()) {
+      ConnectivityStateInfo stateInfo = getSubchannelStateInfoRef(subchannel).get();
+      overallState = aggregateState(overallState, stateInfo.getState());
+    }
+    if (overallState != currentState || picker != currentPicker || subchannelEnteredReady) {
+      helper.updateBalancingState(overallState, picker);
+      currentState = overallState;
+      currentPicker = picker;
+      subchannelEnteredReady = false;
+    }
+  }
+
+  private void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateInfo) {
+    if (subchannels.get(stripAttrs(subchannel.getAddresses())) != subchannel) {
+      return;
+    }
+    AtomicReference<ConnectivityStateInfo> subchannelStateRef =
+        getSubchannelStateInfoRef(subchannel);
+
+    // Don't proactively reconnect if the subchannel enters IDLE, even if previously was connected.
+    // If the subchannel was previously in TRANSIENT_FAILURE, it is considered to stay in
+    // TRANSIENT_FAILURE until it becomes READY.
+    if (stateInfo.getState().equals(READY)) {
+      subchannelEnteredReady = true;
```

Comment:
Why use a flag? `updateBalancingState(currentPicker)` is called immediately after this and then set this flag to false. There is no difference if not setting `subchannelEnteredReady` at all.

Reply:
`updateBalancingState(...)` may not necessarily trigger Channel reprocess. It intends to only trigger it when 

a) the overall balancing state has changed; or
b) the ring/backend list has changed; or
c) any subchannels enters READY

So it will not trigger Channel reprocess when 6 out of 10 subchannels become CONNECTING (the rests are in IDLE) while previously there were 5 out 10 in CONNECTING.

Note, this will probably be changed after I update the implementation according to the latest update in the design. Stay tuned.

Reply:
Without introducing `subchannelEnteredReady` at all,  `updateBalancingState(SubchannelPicker picke)` can still do only trigger repocess when

a) the overall balancing state has changed; or
b) the ring/backend list has changed; or
c) any subchannel enters READY from (non-READY overall state)

The only difference for introducing `subchannelEnteredReady` is when any subchannel enters READY from _READY_ overall state, it will call `helper.updateBalancingState(READY, currentPicker)` again, which is not necessary.

Reply:
> The only difference for introducing subchannelEnteredReady is when any subchannel enters READY from READY overall state, it will call helper.updateBalancingState(READY, currentPicker) again.

That's the point. Invoking Channel reprocessing is indeed necessary in such cases. For example, if previously subchannels S1 was in READY while S2 was in CONNECTING, RPCs hashed to S2 would have been buffered. When S2 enters READY, we need to trigger a Channel reprocess to make those PRCs pick subchannel again. Otherwise, they will be buffered indefinitely even though the subchannel hashed for them has been READY.

Again, as I mentioned, this behavior will probably be changed based on the latest design update. Everything mentioned here may no longer apply.

Reply:
Using the same RingHashPicker reference for different `updateBalancingState()` invocations has serious problem. The RingHashPicker is mutable, `pickSubchannel()` is not thread-safe. Why not make RingHashPicker immutable by snapshotting the stateInfos for all subchannels? Then no longer need `subchannelEnteredReady`, just picker comparison will work.  Okay, I will stay tuned.

Reply:
Obsolete